gateway/src/main/java/io/camunda/zeebe/gateway/admin/backup/BackupRequestHandler.java

```diff
@@ -25,40 +31,144 @@ public BackupRequestHandler(final BrokerClient brokerClient) {
-                  "Expected to send request to all %d partitions, but found only %d partitions in topology."
-                      .formatted(expectedPartitionCount, knownPartitions))));
+          operationFailed("take", backupId, topologyComplete.getLeft()));
     }
 
     final var backupTriggered =
-        topology.getPartitions().stream()
-            .map(partitionId -> getRequestForPartition(backupId, partitionId))
+        topologyManager.getTopology().getPartitions().stream()
+            .map(partitionId -> getBackupRequestForPartition(backupId, partitionId))
             .map(brokerClient::sendRequestWithRetry)
             .toArray(CompletableFuture[]::new);
 
     return CompletableFuture.allOf(backupTriggered)
         .thenApply(ignore -> backupId)
         .exceptionallyCompose(
-            error -> CompletableFuture.failedFuture(backupFailed(backupId, error.getCause())));
+            error ->
+                CompletableFuture.failedFuture(
+                    operationFailed("take", backupId, error.getCause())));
+  }
+
+  @Override
+  public CompletableFuture<BackupStatus> getStatus(final long backupId) {
+    final Either<Throwable, Boolean> topologyComplete = checkTopologyComplete();
+    if (topologyComplete.isLeft()) {
+      return CompletableFuture.failedFuture(
+          operationFailed("query", backupId, topologyComplete.getLeft()));
+    }
+
+    final var statusesReceived =
+        topologyManager.getTopology().getPartitions().stream()
+            .map(partitionId -> getStatusQueryForPartition(backupId, partitionId))
+            .map(brokerClient::sendRequestWithRetry)
+            .toList();
+
+    return CompletableFuture.allOf(statusesReceived.toArray(CompletableFuture[]::new))
+        .thenApply(ignore -> aggregatePartitionStatus(backupId, statusesReceived))
+        .exceptionallyCompose(
+            error ->
+                CompletableFuture.failedFuture(
+                    operationFailed("query", backupId, error.getCause())));
+  }
+
+  private Either<Throwable, Boolean> checkTopologyComplete() {
+    final BrokerClusterState topology = topologyManager.getTopology();
+    if (topology == null) {
+      return Either.left(new NoTopologyAvailableException());
+    }
+
+    final int expectedPartitionCount = topology.getPartitionsCount();
+    final int knownPartitions = topology.getPartitions().size();
+    if (expectedPartitionCount != knownPartitions) {
+      return Either.left(
+          new IncompleteTopologyException(
+              "Expected to send request to all %d partitions, but found only %d partitions in topology."
+                  .formatted(expectedPartitionCount, knownPartitions)));
+    }
+
+    return Either.right(true);
+  }
+
+  private static BackupOperationFailedException operationFailed(
+      final String operation, final long backupId, final Throwable error) {
+    return new BackupOperationFailedException(operation, backupId, error);
```

Comment:
:question: I'm not sure I get why we abstracted a single constructor call away into a method with the same arguments/no additional logic. Is this a leftover from previous approaches? Or just to ensure that we always fail with the same exception (in which case, it could still happen that someone forgets to call `operationFailed` anyway)?